ion-schema/src/isl/mod.rs

```diff
@@ -173,48 +183,203 @@ impl Display for IslVersion {
     }
 }
 
-/// Provides a method to write an ISL model using an Ion writer
-pub trait WriteToIsl {
-    fn write_to<W: IonWriter>(&self, writer: &mut W) -> IonSchemaResult<()>;
+/// Models the content that could be in a Schema document.
+#[derive(Debug, Clone, PartialEq)]
+pub enum SchemaContent {
+    Version(IslVersion),
+    Header(SchemaHeader),
+    Type(IslType),
+    Footer(SchemaFooter),
+    OpenContent(Element),
+}
+impl From<IslVersion> for SchemaContent {
+    fn from(value: IslVersion) -> Self {
+        SchemaContent::Version(value)
+    }
+}
+impl From<SchemaHeader> for SchemaContent {
+    fn from(value: SchemaHeader) -> Self {
+        SchemaContent::Header(value)
+    }
+}
+impl From<IslType> for SchemaContent {
+    fn from(value: IslType) -> Self {
+        SchemaContent::Type(value)
+    }
+}
+impl From<SchemaFooter> for SchemaContent {
+    fn from(value: SchemaFooter) -> Self {
+        SchemaContent::Footer(value)
+    }
 }
 
-/// Provides an internal representation of an schema file
-#[derive(Debug, Clone, PartialEq)]
-pub struct IslSchema {
-    /// Represents an id for the given ISL model
-    id: String,
-    /// Represents the ISL version for given schema
-    version: IslVersion,
+impl SchemaContent {
+    fn as_header(&self) -> Option<&SchemaHeader> {
+        if let SchemaContent::Header(schema_header) = self {
+            Some(schema_header)
+        } else {
+            None
+        }
+    }
+    fn as_type(&self) -> Option<&IslType> {
+        if let SchemaContent::Type(value) = self {
+            Some(value)
+        } else {
+            None
+        }
+    }
+    fn as_open_content(&self) -> Option<&Element> {
+        if let SchemaContent::OpenContent(value) = self {
+            Some(value)
+        } else {
+            None
+        }
+    }
+    fn expect_header(&self) -> &SchemaHeader {
+        if let SchemaContent::Header(value) = self {
+            value
+        } else {
+            panic!("illegal state encountered; this should be unreachable. Expected to find a Header, but found: {:?}", self)
+        }
```

Comment:
For all of these, you could do:
```suggestion
        let SchemaContent::Header(value) = self else {
            unreachable!("expected to find a Header, but found: {:?}", self)
        };
        value
```